src/providers/aws.rs

```diff
@@ -154,75 +233,6 @@ pub struct AWSRegion {
     log: Option<slog::Logger>,
 }
 
-impl super::Launcher for AWSRegion {
-    type Region = String;
-    type Machine = MachineSetup;
-
-    fn region(&self) -> Self::Region {
-        self.region.name().to_string()
-    }
-
-    fn launch(
-        &mut self,
-        l: super::LaunchDescriptor<Self::Machine, Self::Region>,
-    ) -> Result<(), Error> {
-        self.log = Some(l.log);
-        self.make_spot_instance_requests(
-            l.max_instance_duration
-                .map(|x| (x.as_secs() / 60) as i64)
-                .unwrap_or_else(|| 360),
-            l.machines,
-        )?;
-
-        let start = time::Instant::now();
-        self.wait_for_spot_instance_requests(l.max_wait)?;
-        if let Some(mut d) = l.max_wait {
-            d -= time::Instant::now().duration_since(start);
-        }
-
-        self.wait_for_instances(l.max_wait)
-    }
-
-    fn connect_instances<'l>(&'l self) -> Result<HashMap<String, crate::Machine<'l>>, Error> {
-        let log = self.log.as_ref().unwrap();
-        let private_key_path = self.private_key_path.as_ref().unwrap();
-        self.instances
-            .values()
-            .map(|info| match info {
-                (Some((public_ip, public_dns)), (name, MachineSetup { .. })) => {
-                    use std::net::{IpAddr, SocketAddr};
-                    let sess = ssh::Session::connect(
-                        log,
-                        "ubuntu",
-                        SocketAddr::new(
-                            public_ip
-                                .parse::<IpAddr>()
-                                .context("machine ip is not an ip address")?,
-                            22,
-                        ),
-                        Some(private_key_path.path()),
-                        None,
-                    )
-                    .context(format!("failed to ssh to machine {}", public_dns))
-                    .map_err(|e| {
-                        error!(log, "failed to ssh to {}", public_ip);
-                        e
-                    })?;
-                    let machine = Machine {
-                        public_ip: public_ip.clone(),
-                        public_dns: public_dns.clone(),
-                        nickname: name.clone(),
-                        ssh: Some(sess),
-                        _tsunami: Default::default(),
-                    };
-                    Ok((name.clone(), machine))
-                }
-                _ => bail!("Machines not initialized"),
-            })
-            .collect()
-    }
-}
-
 impl AWSRegion {
     /// Connect to AWS region `region`, using credentials provider `provider`.
     pub fn new<P>(region: &str, provider: P, log: slog::Logger) -> Result<Self, Error>
```

Comment:
This probably doesn't need to be `pub` anymore?

Reply:
hm, I sort of like the idea of leaving the lower-level API exposed in case `TsunamiBuilder` doesn't work for some use case. So while it doesn't *need* to be `pub`, there's no harm in making it `pub` - although the pub methods should have documentation explaining what they do.